crates/hir-def/src/path.rs

```diff
@@ -47,20 +48,30 @@ impl Display for ImportAliasDisplay<'_> {
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub enum Path {
-    /// A normal path
-    Normal {
-        /// Type based path like `<T>::foo`.
-        /// Note that paths like `<Type as Trait>::foo` are desugared to `Trait::<Self=Type>::foo`.
-        type_anchor: Option<Interned<TypeRef>>,
-        mod_path: Interned<ModPath>,
-        /// Invariant: the same len as `self.mod_path.segments` or `None` if all segments are `None`.
-        generic_args: Option<Box<[Option<Interned<GenericArgs>>]>>,
-    },
+    BarePath(Interned<ModPath>),
+    /// You can **not** rely on `Path::Normal` to not have generic args and type anchor, it may be missing both
```

Comment:
Honestly I'm split about this at all, I'm not sure not preserving the invariant is worth it. I wrote it because of `qualifier()` that can break the invariant, and it mainly affects `is_self_type()` that needs to adjust when the variant is not kept, but it is easy enough to do the opposite and prefer `is_self_type()`.